src/pyj/read_book/cfi.pyj

```diff
@@ -536,153 +543,155 @@ def at(x, y, doc):  # {{{
+#
+# Includes everything that the decode() function does, in addition to:
+# range: A text range, as desribed above.
+# use_range_end_pos: If this is True, a position calculated from the range should
+#                    use the position after the last character in the range.
+#                    (This is set if the offset is equal to the length of the text in the node.)
+def decode_with_range(cfi, doc):  # {{{
     doc = doc or window.document
-    r = decode(cfi, doc)
-    if not r:
+    decoded = decode(cfi, doc)
+    if not decoded:
         return None
-    node = r.node
+    node = decoded.node
     ndoc = node.ownerDocument
     if not ndoc:
         print(str.format("CFI node has no owner document: {} {}", cfi, node))
         return None
 
-    x = None
-    y = None
     range_ = None
+    position_at_end_of_range = None
 
-    if jstype(r.offset) is "number":
-        # Character offset
-        range_ = ndoc.createRange()
-        if r.forward:
-            try_list = [{'start':0, 'end':0, 'a':0.5}, {'start':0, 'end':1, 'a':1}, {'start':-1, 'end':0, 'a':0}]
-        else:
-            try_list = [{'start':0, 'end':0, 'a':0.5}, {'start':-1, 'end':0, 'a':0}, {'start':0, 'end':1, 'a':1}]
-        a = None
+    if jstype(decoded.offset) is "number":
+        # We can only create a meaningful range if the node length is
+        # positive and nonzero.
         node_len = node.nodeValue.length if node.nodeValue else 0
-        offset = r.offset
-        if not offset:
-            range_.setStart(node, 0)
-            range_.setEnd(node, 0)
-        else:
-            rects = v'[]'
-            for v'var i = 0; i < 2; i++':
-                # Try reducing the offset by 1 if we get no match as if it refers to the position after the
-                # last character we wont get a match with getClientRects
-                offset = r.offset - i
-                if offset < 0:
-                    offset = 0
-                k = 0
-                while not rects?.length and k < try_list.length:
-                    t = try_list[k]
-                    k += 1
-                    start_offset = offset + t.start
-                    end_offset = offset + t.end
-                    a = t.a
-                    if start_offset < 0 or end_offset >= node_len:
-                        continue
-                    range_.setStart(node, start_offset)
-                    range_.setEnd(node, end_offset)
-                    rects = range_.getClientRects()
+        if node_len:
+            range_ = ndoc.createRange()
+
+            # Check for special case: End of range offset, after the last character
+            offset = decoded.offset
+            position_at_end_of_range = False
+            if offset == node_len:
+                offset -= 1
+                position_at_end_of_range = True
+
+            range_.setStart(node, offset)
+            range_.setEnd(node, offset + 1)
```

Comment:
This is important. The "if not offset" code was causing a collapsed range to be created, which doesn't have a valid bounding box. I changed it to always create a range of length 1 starting at the offset (or the offset minus one in the case the offset is equal to the length, with the flag to indicate this.)